### ISSUE TITLE
Use pluggable deferred/promise everywhere

### ISSUE THREAD

**Issue by oberstet** [crossbario/autobahn-js#512]
ABJS allows to switch the deferred/promise implementation to be used between when.js and native builtin:

https://github.com/crossbario/autobahn-js/blob/e4e59136f5e25b49007665dfa0ab2cb6f9e1d8d6/packages/autobahn/lib/util.js#L364

this can then be used like so:

https://github.com/crossbario/autobahn-js/blob/e4e59136f5e25b49007665dfa0ab2cb6f9e1d8d6/packages/autobahn/lib/connection.js#L31

however, this pluggability isn't currently used throughout, eg.

```
(cpy382_1) oberstet@intel-nuci7:~/scm/crossbario/autobahn-js$ find packages/autobahn/lib -name "*.js" -exec grep -Hi "when\." {} \;
packages/autobahn/lib/util.js:   var d = when.defer();
packages/autobahn/lib/util.js:   var d = when.defer();
packages/autobahn/lib/util.js:         defer = when.defer;
packages/autobahn/lib/util.js:      defer = when.defer;
(cpy382_1) oberstet@intel-nuci7:~/scm/crossbario/autobahn-js$ find packages/autobahn/lib -name "*.js" -exec grep -Hi "when_fn" {} \;
packages/autobahn/lib/auth/persona.js:var when_fn = require("when/function");
packages/autobahn/lib/session.js:var when_fn = require("when/function");
packages/autobahn/lib/session.js:         when_fn.call(reg.endpoint, args, kwargs, cd).then(
packages/autobahn/lib/session.js:               when_fn.call(self._onchallenge, self, method, extra).then(
(cpy382_1) oberstet@intel-nuci7:~/scm/crossbario/autobahn-js$ find packages/autobahn-xbr/lib -name "*.js" -exec grep -Hi "when\." {} \;
packages/autobahn-xbr/lib/seller.js:                    var d1 = when.all(pl1).then(
packages/autobahn-xbr/lib/seller.js:                    var d2 = when.all(pl2).then(
packages/autobahn-xbr/lib/seller.js:                    when.all([d1, d2]).then(
packages/autobahn-xbr/lib/util.js:            defer = when.defer;
packages/autobahn-xbr/lib/util.js:        defer = when.defer;
(cpy382_1) oberstet@intel-nuci7:~/scm/crossbario/autobahn-js$ find packages/autobahn-xbr/lib -name "*.js" -exec grep -Hi "when_fn" {} \;
(cpy382_1) oberstet@intel-nuci7:~/scm/crossbario/autobahn-js$ 
```

---

one issue: when.js includes helpers like when.all, which at least historically haven't been avail in native impl. .. we should recheck this. if it is still not there, then no choice: continue using when.js (for those helpers) ..

**Comment by om26er** [comment]
Using https://github.com/crossbario/autobahn-js/blob/e4e59136f5e25b49007665dfa0ab2cb6f9e1d8d6/packages/autobahn/lib/connection.js#L31 is no problem. However we need to be able to find a way to deal with the use of `when_fn = require("when/function");`. For that we need to write a polyfill, or figure out if that functionality exists somewhere

**Comment by om26er** [comment]
After looking a bit into this, I believe we should remove the use of whenjs all together and use ES6 Promise by default (we no longer support ES5 anyways). `whenjs` could just be supported optionally by explicitly providing deferred factory by the user, like https://github.com/crossbario/autobahn-js/blob/master/packages/autobahn/lib/util.js#L391

@oberstet Is there something I am missing here ?

**Comment by oberstet** [comment]
> I believe we should remove the use of whenjs all together and use ES6 Promise by default

In general, yes, but probably wait for TC39.

Background: there are 2 elements that are oftem needed in practice, IMO (both in libraries and apps): **all()** and **any**

see: https://github.com/tc39/proposal-promise-any#motivation

at least **all** is used in XBR seller:

```console
oberstet@intel-nuci7:~$ find scm/crossbario/autobahn-js/packages/autobahn/lib/ -name "*.js" -exec grep -Hi "when\.all" {} \;
oberstet@intel-nuci7:~$ find scm/crossbario/autobahn-js/packages/autobahn-xbr/lib/ -name "*.js" -exec grep -Hi "when\.all" {} \;
scm/crossbario/autobahn-js/packages/autobahn-xbr/lib/seller.js:                    var d1 = when.all(pl1).then(
scm/crossbario/autobahn-js/packages/autobahn-xbr/lib/seller.js:                    var d2 = when.all(pl2).then(
scm/crossbario/autobahn-js/packages/autobahn-xbr/lib/seller.js:                    when.all([d1, d2]).then(
oberstet@intel-nuci7:~$ find scm/crossbario/autobahn-js/packages/autobahn/lib/ -name "*.js" -exec grep -Hi "when\.any" {} \;
oberstet@intel-nuci7:~$ find scm/crossbario/autobahn-js/packages/autobahn-xbr/lib/ -name "*.js" -exec grep -Hi "when\.any" {} \;
```
# all

supported in ECMAScript 2015 ("ES2015") and all major browsers/-runtimes

* https://github.com/cujojs/when/blob/master/docs/api.md#whenall
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all
* https://tc39.es/ecma262/#sec-promise.all

# any

only proposed in TC39, not supported by browsers (not sure about Node?)

* https://github.com/cujojs/when/blob/master/docs/api.md#whenany
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any
* https://github.com/tc39/proposal-promise-any

**Comment by oberstet** [comment]
IOW: I would keep when.js for now, and file this as a separate issue: "make ES2021 promise default and when.js optional" - once TC39 has landed

**Comment by om26er** [comment]
I took a stab at that here https://github.com/crossbario/autobahn-js/pull/516

**Comment by oberstet** [comment]
fixed via #516
